lib/wal/wal_replay.c

```diff
@@ -578,6 +578,27 @@ wal_replay_gen_impl(struct wal_replay *rep, struct wal_replay_gen *rgen, bool fl
  * General WAL replay interfaces
  */
 
+/*
+ * WAL force replay:
+ *
+ * The objective of "durability.replay.force" KVDB rparam is to instruct WAL to replay as
+ * much data as possible and bring the kvdb back to a consistent state.
+ *
+ * For instance, if the app fails due to a ENOSPC health event, then it is highly likely
+ * that the replay would also fail with ENOSPC when it re-ingests the same data.
+ *
+ * The user has two options here to bring the affected KVDB online:
+ *
+ * 1. Extend the file-system such that the affected KVDB has enough free space for the
+ *    replay to succeed
+ *
+ * 2. Open the affected KVDB in force replay mode. In this mode, WAL replays as much data
+ *    as possible before bringing the KVDB back into a consistent state. The user must be
+ *    aware that there can be data loss when opening a KVDB in this mode.
+ *
+ * TODO: Review other places in the wal replay code that can make progress without
+ *       failing if the force replay flag is set
```

Comment:
So much to discuss:
- I would avoid the use of "consistent state" in the first paragraph and in bullet #2.  It's not a well defined term in this context and will mean different things to different people.
- Does corruption or data loss occur when you try to open w/ replay enabled and it cannot complete due to ENOSPC errors?  Or does it just fail with data still left in the WAL?
- I don't think it's a good idea to provide a "force open but data loss is okay" option.  There may be other ways to get back to a WAL-clean state. For example, I bet "hse kvdb compact" can be run safely even with data in WAL. If so, we could advise an offline compaction, then re-open.

Reply:
>>>> Does corruption or data loss occur when you try to open w/ replay enabled and it cannot complete due to ENOSPC errors? Or does it just fail with data still left in the WAL?

It fails with data still left in WAL.

>>>> I don't think it's a good idea to provide a "force open but data loss is okay" option. There may be other ways to get back to a WAL-clean state.

Agreed that there might be other ways and the force replay must be used as a last resort. Btw, having different recovery modes is not something new, RockDB supports something similar - https://github.com/facebook/rocksdb/wiki/WAL-Recovery-Modes

Reply:
Much better wording.